File: pfm-sdk/src/main/java/com/tink/pfmsdk/mapper/StatisticToPeriodBalanceConverter.java
package com.tink.pfmsdk.mapper;

import com.tink.pfmsdk.charts.models.PeriodBalance;
import se.tink.core.models.misc.Period;
import se.tink.core.models.statistic.Statistic;
import se.tink.modelConverter.AbstractConverter;

class StatisticToPeriodBalanceConverter extends AbstractConverter<Statistic, PeriodBalance> {

	private ModelConverterImplementation modelConverter;

	public StatisticToPeriodBalanceConverter(ModelConverterImplementation modelConverter) {
		this.modelConverter = modelConverter;
	}

	@Override
	public PeriodBalance convert(Statistic source) {
		PeriodBalance pb = new PeriodBalance();
		if (source.getAmount().getValue() != null) {
			pb.setAmount(source.getAmount().getValue().doubleValue());
		}
		if (source.getPeriod() != null) {
			pb.setPeriod(modelConverter.map(source.getPeriod(), Period.class));
		}
		return pb;
	}

	@Override
	public Class<Statistic> getSourceClass() {
		return Statistic.class;
	}

	@Override
	public Class<PeriodBalance> getDestinationClass() {
		return PeriodBalance.class;
	}
}

File: pfm-ui/src/main/java/com/tink/pfmui/mapper/StatisticToPeriodBalanceConverter.kt
package com.tink.pfmui.mapper

import com.tink.pfmui.charts.models.PeriodBalance
import se.tink.commons.extensions.doubleValue
import com.tink.model.time.Period
import com.tink.model.statistic.Statistic
import se.tink.modelConverter.AbstractConverter

internal class StatisticToPeriodBalanceConverter(
    private val modelConverter: ModelConverterImplementation
) : AbstractConverter<Statistic, PeriodBalance>() {

    override fun convert(source: Statistic): PeriodBalance = source.toPeriodBalance()

    private fun Statistic.toPeriodBalance() =
        PeriodBalance(
            period.let { modelConverter.map(it, Period::class.java) },
            value.value.doubleValue()
        )

    override fun getSourceClass(): Class<Statistic> {
        return Statistic::class.java
    }

    override fun getDestinationClass(): Class<PeriodBalance> {
        return PeriodBalance::class.java
    }
}